### PR TITLE
Alice B and Anna

### PR DESCRIPTION
**Pull request by annakim93**
# Assignment Submission: Rideshare Rails

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions. These questions should be answered by all team members together, not by a single teammate.

## Reflection

Prompt | Response
--- | ---
Describe the types of entity relationships you set up in your project and why you set up the relationships that way | Trips belong to Passengers and belongs to Drivers; Passenger has many trips; Driver has many trips; we did it this way because a trip has a passenger and a driver and passengers and drivers can take many trips
Describe the role of model validations in your application | Make sure we have the required fields for the drivers, passengers, and to create a trip
How did your team break up the work to be done? | We split up passenger and driver and trips and did the tests for each one
What features did you choose to prioritize in your project, and what features, if any, did you have to set aside to meet the deadline? | We prioritized having all the logical model methods done and having a good looking site that works! 
What was one thing that your team collectively gained more clarity on after completing this assignment? | How to use Rails!
What are two discussion points that you and your team discussed when giving/receiving feedback from each other that you would be willing to share? | We discussed how we like to work - this was our first time splitting up the work and pulling and pushing to Git; we learned to communicate a lot so that we can do our own parts and still come together and resolve merge conflicts
Optional: What is the URL of your deployed Heroku app? |

### CODE REVIEW

**Comment by aliceboone** [comment]
https://enjotheview.herokuapp.com/

File: app/models/trip.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  def self.assign_driver
+    if Driver.find_by(available: 'true').nil?
+      return nil
+    else
+      return Driver.find_by(available: 'true').id
+    end
+  end
+
+  def change_driver_status
```
Comment (beccaelenzil) [inline app/models/trip.rb:23]: Great use of a custom model method. Consider creating this as a Driver class method.

File: app/views/drivers/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      <div class="col-75">
+        <%= f.text_field :vin %>
+      </div>
+    </div>
+      <div class="row">
+      <div class="col-25">
+        <%= f.label :available %>
+      </div>
+      <div class="col-75">
+        <%= f.text_field :available %>
```
Comment (beccaelenzil) [inline app/views/drivers/_form.html.erb:35]: Consider whether you want the use to be able to enter the status of a driver. It could default to true and could only be changed using your custom action. It is not intuitive what should be entered in this form field.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  root to: 'homepages#index'
+
+  resources :drivers do
+    resources :trips, only: [:index]
+  end
+
+  patch 'drivers/:id/toggle_availability', to: 'drivers#toggle_availability', as: :toggle_availability
```
Comment (beccaelenzil) [inline config/routes.rb:8]: Great use of custom route.

File: test/models/driver_test.rb [hunk 4]
```diff
@@ -60,22 +72,40 @@
 
   # Tests for methods you create should go here
   describe "custom methods" do
-    describe "average rating" do
-      # Your code here
+    it "average rating: will return nil if there were no trips or ratings recorded" do
```
Comment (beccaelenzil) [inline test/models/driver_test.rb:75]: Good work testing this edge case

File: app/views/passengers/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+              <button><%= link_to "Delete", passenger_path(@passenger.id), method: :delete, data: { confirm: "Please confirm deletion." } %></button>
+              <button><%= link_to "Back", passengers_path %></button>
+         </li>
+          <li><button class="request_trip"><%= link_to "Request Ride", passenger_trips_path(@passenger.id), method: :post %></button></li>
+      </ul>
+  </div>
+
+<section>
+  <h3><%= @passenger.name %>'s trips</h3>
+  <%= render partial: "trips" %>
```
Comment (beccaelenzil) [inline app/views/passengers/show.html.erb:22]: Nice use of a partial view.

**Review by beccaelenzil — COMMENTED** [review]
# Rideshare Rails



## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| Practices collaborating with git, and all team members contribute git commits and best git practices | ✔️
| Demonstrates understanding of relationships by giving accurate answers to the reflection questions and appropriate code in the models | ✔️
| Logic to calculate a driver's average rating and total earnings is located in the model, and it has unit tests | ✔️
| There are reasonable tests to test the validation requirements for all models | ✔️
| There are reasonable tests to test the relationship requirements for all models | ✔️
| There are reasonable tests to test the controller actions for all controllers | ✔️
| The app has an attractive and usable user interface | ✔️ Nice user experience!
| The app uses/is compatible with database seeds | ✔️
| Router code is clean: uses `resources` and RESTful routes | ✔️ Good work limiting routes

## Functional Requirements

| Functional Requirement | yes/no |
| --- | --- |
| On the passenger's details page, I want to be able to see total charged, list of trips, a link to edit, and delete | ✔️
| When adding a new passenger, I want to see errors and validations that show that a passenger must be provided a name and a phone number, so that I cannot make a passenger without name or phone number | ✔️
| On the passenger's details page, I can create a new trip for this passenger, with an assigned driver and no rating | ✔️
| On the driver's details page, I want to be able to see total earnings, average rating, list of trips, a link to edit, and delete | ✔️
| When adding a new driver, I want to see errors and validations that show that a driver must be provided a name and VIN, so that I cannot make a driver without name or VIN | ✔️
| On the trip's detail page, I want to be able to view details, assign a rating, navigate to the trip's passenger, driver, a link to edit, and delete | ✔️

## Overall Feedback

Excellent job overall. This code is well-written and well-tested. It is clear that the learning goals around model relationships, validations, and seeding were met. You've created a great user experience with attractive styles. Furthermore, great work using `Flash` and parital views. Keep up the hard work!

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 8+ in Code Review && 5+ in Functional Requirements | ✔️
| Yellow (Approaches Standards) | 5+ in Code Review && 4+ in Functional Requirements, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-4 in Code Review or 0-3 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
